nomad/auth/auth.go

```diff
@@ -0,0 +1,483 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package auth
+
+import (
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	metrics "github.com/armon/go-metrics"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad/acl"
+	"github.com/hashicorp/nomad/helper"
+	"github.com/hashicorp/nomad/nomad/state"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// aclCacheSize is the number of ACL objects to keep cached. ACLs have a parsing
+// and construction cost, so we keep the hot objects cached to reduce the ACL
+// token resolution time.
+const aclCacheSize = 512
+
+type StateGetter func() *state.StateStore
+type LeaderACLGetter func() string
+
+type RPCContext interface {
+	IsTLS() bool
+	IsStatic() bool
+	Certificate() *x509.Certificate
+	GetRemoteIP() (net.IP, error)
+}
+
+type Encrypter interface {
+	VerifyClaim(string) (*structs.IdentityClaims, error)
+}
+
+type Authenticator struct {
+	aclsEnabled  bool
+	tlsEnabled   bool
+	logger       hclog.Logger
+	getState     StateGetter
+	getLeaderACL LeaderACLGetter
+	region       string
+
+	// aclCache is used to maintain the parsed ACL objects
+	aclCache *structs.ACLCache[*acl.ACL]
+
+	// encrypter is a pointer to the server's Encrypter that can be used to
+	// verify claims
+	encrypter Encrypter
+}
+
+func NewAuthenticator(stateFn StateGetter, logger hclog.Logger, getLeaderACLFn LeaderACLGetter, aclsEnabled, tlsEnabled bool, region string, encrypter Encrypter,
```

Comment:
Would there be a benefit to wrapping the function parameters in a struct? I am thinking if we add any in the future, the function signature will get a little unwieldily.

Reply:
Can't hurt. It's only instantiated once per Nomad server, so it's not like that allocation is in a hot path. Will do.

Reply:
Done!